File: aoc/2024/aoc202203.hs
module AOC202203 where

import Data.Char(ord)
import Data.Set(elems,fromList,intersection)

import AOC

aoc = AOC {
    day="../../2022/input/03",
    aocTests=[
        AOCTest {
            testData=unlines [
                "vJrwpWtwJgWrhcsFMMfFFhFp",
                "jqHRNqRjqzjGDLGLrsFMfFZSrLrFZsSL",
                "PmmdzqPrVvPwwTWBwg",
                "wMqvLMZHhHMvwLHjbvcjnnSBnvTQFn",
                "ttgJtRGJQctTZtZT",
                "CrZsJsPPZsGzwwsLwLmpwMDw"
                ],
            testResult=Just "157",
            testResult2=Just "70"
            }
        ],
    aocCode=Code {
        codeParse=parse,
        codeParse2=parse2,
        codeTest=result,
        codeTest2=result2,
        codeResult=result,
        codeResult2=result2
        }
    }

parse = map p . lines
  where p l = splitAt (length l `div` 2) l

priority :: Char -> Int
priority ch
  | ch >= 'a' && ch <= 'z' = ord ch - ord 'a' + 1
  | ch >= 'A' && ch <= 'Z' = ord ch - ord 'A' + 27

common :: (String,String) -> Char
common (a,b) = minimum $ fromList a `intersection` fromList b

result = sum . map (priority . common)

parse2 = p . lines
  where p l | null this = [] | otherwise = this : p rest
          where (this,rest) = splitAt 3 l

badge :: [String] -> Char
badge [a,b,c] =
    minimum $ fromList a `intersection` fromList b `intersection` fromList c

result2 = sum . map (priority . badge)

File: aoc/2025/aoc202203.go
package main

func init() {
	Register(&aoc202203{
		AOC: AOC{
			Day:           3,
			InputFilename: "../2022/input/03.txt",
			Tests: []Test{
				Test{
					Input: `vJrwpWtwJgWrhcsFMMfFFhFp
jqHRNqRjqzjGDLGLrsFMfFZSrLrFZsSL
PmmdzqPrVvPwwTWBwg
wMqvLMZHhHMvwLHjbvcjnnSBnvTQFn
ttgJtRGJQctTZtZT
CrZsJsPPZsGzwwsLwLmpwMDw
`,
					Part1: "157",
					Part2: "70",
				},
			},
		},
	})
}

type aoc202203 struct {
	AOC
}

func (aoc *aoc202203) c2p(ch byte) int {
	if ch >= 'a' {
		return int(ch - 'a' + 1)
	} else {
		return int(ch - 'A' + 27)
	}
}

func (aoc *aoc202203) s2p(set uint64) int {
	for i := range 64 {
		if set == 1 {
			return i
		}
		set >>= 1
	}
	panic("bad input")
}

func (aoc *aoc202203) Part1(input *Input) string {
	result := 0
	for line := range input.Lines() {
		n := len(line)
		var a, b uint64
		for i, l2 := range []byte(line[n/2:]) {
			a |= 1 << aoc.c2p(line[i])
			b |= 1 << aoc.c2p(l2)
		}
		result += aoc.s2p(a & b)
	}
	return IntResult(result)
}

func (aoc *aoc202203) next(input *Input) (uint64, bool) {
	line, ok := input.Line()
	if !ok {
		return 0, false
	}
	var sack uint64
	for _, ch := range []byte(line) {
		sack |= 1 << aoc.c2p(ch)
	}
	return sack, true
}

func (aoc *aoc202203) Part2(input *Input) string {
	result := 0
	for {
		a, _ := aoc.next(input)
		b, _ := aoc.next(input)
		c, ok := aoc.next(input)
		if !ok {
			break
		}
		result += aoc.s2p(a & b & c)
	}
	return IntResult(result)
}
